common/core/desktop/src/kmx/kmx_processor.cpp

```diff
@@ -240,28 +249,43 @@ KMX_BOOL KMX_Processor::ProcessGroup(LPGROUP gp, KMX_BOOL *pOutputKeystroke)
         PKMX_WCHAR pdeletecontext = m_context.Buf(1);   // I4933
         if(!pdeletecontext || *pdeletecontext == 0) {   // I4933
           m_actions.QueueAction(QIT_INVALIDATECONTEXT, 0);
+          if(m_debug_items) {
+            m_debug_items->push_group_exit(gp, KM_KBP_DEBUG_FLAG_NOMATCH);
+          }
           *pOutputKeystroke = TRUE;   // I4933
           return FALSE;   // I4933
         }
         m_actions.QueueAction(QIT_BACK, BK_BACKSPACE);   // I4933
       } else {   // I4024   // I4128   // I4287   // I4290
         DebugLog(" ... IsLegacy = FALSE; IsTIP = TRUE");   // I4128
         m_actions.QueueAction(QIT_INVALIDATECONTEXT, 0);
+        if(m_debug_items) {
+          m_debug_items->push_group_exit(gp, KM_KBP_DEBUG_FLAG_NOMATCH);
+        }
         *pOutputKeystroke = TRUE;
         return FALSE;
       }
     }
     else if (gp->dpNoMatch != NULL && *gp->dpNoMatch != 0)
     {
       /* NoMatch rule found, and is a character key */
+      if(m_debug_items) {
+        m_debug_items->push_nomatch_enter(gp);
+      }
       PostString(gp->dpNoMatch, m_keyboard.Keyboard, NULL, pOutputKeystroke);
+      if(m_debug_items) {
+        m_debug_items->push_nomatch_exit(gp);
+      }
     }
     else if (m_state.charCode != 0 && m_state.charCode != 0xFFFF && gp->fUsingKeys)
     {
       /* No rule found, is a character key */
       m_actions.QueueAction(QIT_CHAR, m_state.charCode);
     }
 
+    if(m_debug_items) {
+      m_debug_items->push_group_exit(gp, KM_KBP_DEBUG_FLAG_NOMATCH);
+    }
```

Comment:
Corresponds to:
https://github.com/keymanapp/keyman/blob/75e47c1ccc36e1698bafa8178acf8140f88e4e2b/windows/src/engine/keyman32/kmprocess.cpp#L411